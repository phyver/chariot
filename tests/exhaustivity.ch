:set continue_on_error true
:set show_priorities false

data nat where Zero : nat | Succ : nat -> nat

(* INCOMPLETE *)
val f 0 = ???
  | f (n+2) = ???


data list('x) where
    | Nil : list('x)
    | Cons : 'x -> list('x) -> list('x)

:set allow_incomplete_defs true

val last [x] = x
  | last (x::xs) = last xs

:set allow_incomplete_defs false

val last2 [x] = x
  | last2 (x::xs) = last2 xs


val merge f [] ys = ys
  | merge f xs [] = xs
  | merge f (x::xs) (y::ys) = (f x y)::(merge f xs ys)

val test1 [a] = 1
  | test1 [a;b;c] = 2
  | test1 (a::b::c::d::e::l) = 3


-- useless clauses
val test2 x = ???
  | test2 [] = ???
:show test2

val test3 [] = ???
  | test3 ([]::xs) = ???
  | test3 ((y::ys)::xs) = ???
  | test3 _ = ???
:show test3

:set keep_useless_clauses true
val test4 x = ???
  | test4 [] = ???
:show test4

val test5 [] = ???
  | test5 ([]::xs) = ???
  | test5 ((y::ys)::xs) = ???
  | test5 _ = ???
:show test5
