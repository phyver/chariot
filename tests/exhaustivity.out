-- inductive type nat was succesfully defined
-- Typing for f successful
--!! match failures:
--!!   f 1
--!! the definition for f is incomplete
-- inductive type list was succesfully defined
-- Typing for last successful
--!! match failures:
--!!   last []
--!! the definition for last is incomplete
-- Typing for last2 successful
--!! match failures:
--!!   last2 []
--** the definition for last2 is incomplete
-- Typing for merge successful
-- the definition for merge is complete
-- Typing for test1 successful
--!! match failures:
--!!   test1 []
--!!   test1 [x₂; x₄]
--!!   test1 [x₂; x₄; x₆; x₈]
--** the definition for test1 is incomplete
